docs: expand the security page to include more sections (#332)
canonical/jubilant: docs/explanation/security.md

---
myst:
  html_meta:
    description: Security documentation for Jubilant — product architecture, secure by design, cryptography, hardening, logging, decommissioning, security lifecycle, and vulnerability reporting.
---

(security)=

# Security

When producing security documentation for your charm, it's important to consider the security aspects of the charm's tests. If you have any Jubilant-related security questions that aren't answered here, please reach out to the Charm Tech team, and we'll do our best to help.

> ## Product architecture

Reviewer: Do you think we ought to mention anything about secret handling? I guess it's already covered implicitly by this section - unless there's anything unusual we'd want to call out?

Author: I've made it explicit that temporary files are used and that the secret content is not in CLI calls.

Jubilant is a pure-Python library with no persistent state, no network sockets, and no privileged access of its own. Its trust boundary is narrow:

```
test process → jubilant → Juju CLI subprocess → Juju controller
```

- The **test process** calls Jubilant's Python API.
- Jubilant translates each call into one or more {external+juju:ref}`Juju CLI <juju-cli>` invocations via `subprocess`. It does not communicate with Juju or any other service directly over a network.

> - The **Juju CLI** handles all communication with the Juju controller, including authentication, credential management, and TLS. Jubilant never reads, stores, or forwards Juju credentials.

Reviewer: This part could be read as contradicting the susequent paragraph, which mentions `add_credential()`. We say that Jubilant never forwards Juju credentials - I suppose technically this statement excludes cloud credentials, but I still think there's a possibility for confusion. How about an adjustment like below, leading into the following paragraph about passing structured data.

```suggestion
- The **Juju CLI** handles all communication with the Juju controller, including authentication, credential management, and TLS. Jubilant never stores credentials or reads credentials from Juju.
```

> Several Jubilant operations pass structured data to the Juju CLI via temporary files (for example, YAML config passed to `add_cloud()`, `add_credential()`, `run_action()`, and `deploy()`). When the Juju CLI is installed as a snap, these files are written to `~/snap/juju/common/` so the snap can access them; otherwise the system temporary directory is used. In both cases the files are removed immediately after the CLI call returns. Jubilant retains no state between test runs.

Reviewer: `run_action()` should be `run()`. Plus small readability suggestions.

Also, do you think it's worth linking the methods to the corresponding reference docs? Just a thought - I'm not sure it's really worth doing here.

```suggestion
Several Jubilant operations pass structured data to the Juju CLI by writing temporary files. For example, YAML config passed to `add_cloud()`, `add_credential()`, `run()`, and `deploy()`. When the Juju CLI is installed as a snap, these files are written to `~/snap/juju/common/` so the snap can access them; otherwise the system temporary directory is used. In both cases the files are removed immediately after the CLI call returns. Jubilant retains no state between test runs.
```

Author: I worry it would make this noisy, and I don't think people would be wanting to click through to a method from here.

> ## Secure by Design

Reviewer: ```suggestion
## Secure by design
```

Jubilant was deliberately designed as a thin wrapper over the Juju CLI rather than a reimplementation of Juju's protocol. This design choice is a security property: all authentication, credential handling, and network communication is delegated to the Juju CLI, which is independently maintained and audited. Jubilant adds no new network attack surface.

The library does not introduce any new security risks beyond directly running Juju CLI commands. Charm integration tests are most commonly run in ephemeral CI environments. When run in local development environments, Jubilant inherits whatever access the Juju CLI has to that environment — no more.

> See also: {external+juju:ref}`Juju security <juju-security>`

## Cryptographic technology

Jubilant does not use any cryptographic technology, hashing, or digital signatures. All cryptographic operations (TLS, credential storage, API authentication) are handled by the Juju CLI and the Juju controller.

> ## Configuring and operating securely

Reviewer: Can we drop "securely" since the whole topic is security?

```suggestion
## Configuring and operating
```

### Hardening guidelines

> No additional steps are required to harden your integration tests when using Jubilant. Follow the same security and hardening practices as you would if calling the Juju CLI directly.

Reviewer: "No additional steps are required", but we do recommend some good practices. Maybe we should include a link to the "Good practices" section in the "See also" block.

> See also: {external+juju:ref}`Juju | Harden your deployment <harden-your-deployment>`

> ### Logging and monitoring

Reviewer: What if Jubilant crashes for some reason - could sensitive info, such as secrets passed to `add_secret()`, be exposed in the logs?

Author: No, only the filename of the temporary file would get exposed, which is worthless to an attacker (and by the time you see the crash, the temporary file is gone).

Author: I added a note about this.

Jubilant uses Python's standard `logging` module. No security events are logged by Jubilant itself — authentication, authorisation, and access control events are logged by the Juju CLI and the Juju controller, not by the test library.

To capture Jubilant's own log output (command invocations, return codes, and `stderr`), configure a handler for the `jubilant` logger in your test suite:

```python
import logging
logging.getLogger("jubilant").setLevel(logging.DEBUG)
```

For security-relevant events (login failures, credential errors, controller access), consult the Juju controller logs and the Juju CLI's own output rather than Jubilant's logs.

> ## Decommissioning securely

Reviewer: Can we drop "securely" since the whole topic is security?

```suggestion
## Decommissioning
```

> Removing Jubilant from a project requires no special decommissioning steps. Remove `jubilant` from your `pyproject.toml` and re-lock. Jubilant stores no credentials, configuration, or logs of its own. Any Juju environments provisioned during testing should be destroyed using the Juju CLI (`juju destroy-controller`, `juju destroy-model`) independently of removing Jubilant.

Reviewer: Just to be totally clear:

```suggestion
Removing Jubilant from a project requires no special decommissioning steps. Remove `jubilant` from your `pyproject.toml` and re-lock dependencies. Jubilant stores no credentials, configuration, or logs of its own. Any Juju environments provisioned during testing should be destroyed using the Juju CLI (`juju destroy-controller`, `juju destroy-model`) independently of removing Jubilant.
```

## Security lifecycle

Jubilant follows [semantic versioning](https://semver.org/). Security updates are released as patch releases on all major versions that have had releases in the last year. See [SECURITY.md](https://github.com/canonical/jubilant/blob/main/SECURITY.md) for the supported-versions policy.

> **Receiving security updates.** Restrict the version of `jubilant` in `pyproject.toml` in a way that allows picking up new compatible releases every time that you re-lock. For example, `jubilant~=1.2` allows upgrades to 1.3, 1.4, and so on, while staying on the 1.x line.

Reviewer: ```suggestion
**Receiving security updates.** Restrict the version of `jubilant` in `pyproject.toml` in a way that allows picking up new compatible releases every time that you re-lock dependencies. For example, `jubilant~=1.2` allows upgrades to 1.3, 1.4, and so on, while staying on the 1.x line.
```

> **Detecting available updates.** Configure Dependabot or Renovate in your charm repository so that security updates are surfaced automatically as pull requests. Re-lock when prompted so that the charm tests run with the latest version.

Reviewer: ```suggestion
**Detecting available updates.** Configure Dependabot or Renovate in your charm repository so that security updates are surfaced automatically as pull requests. Re-lock dependencies when prompted so that the charm tests run with the latest version.
```

## Reporting vulnerabilities

Report security issues via [GitHub's security advisory for this project](https://github.com/canonical/jubilant/security/advisories/new). See [Privately reporting a security vulnerability](https://docs.github.com/en/code-security/security-advisories/guidance-on-reporting-and-writing/privately-reporting-a-security-vulnerability) for instructions.

You may also send email to security@ubuntu.com. If you want to encrypt your email, follow [Canonical's reporting instructions](https://ubuntu.com/security/disclosure-policy#contact-us).

Known vulnerabilities and past advisories are listed at [github.com/canonical/jubilant/security/advisories](https://github.com/canonical/jubilant/security/advisories).

The [Ubuntu Security disclosure and embargo policy](https://ubuntu.com/security/disclosure-policy) contains more information about what you can expect when you contact us, and what we expect from you.

> ## Good practices

Reviewer: What do you think about adding a good practice related to `cli()`? E.g., do people need to take care with how they escape arguments that are passed to the Juju CLI?

Author: No need to manage escaping, because the subprocess module handles making that safe. But people should avoid args with sensitive data, also when setting config, as they may be in logs. I'll add a note about that.
